### DISCUSSION TITLE
User Authentication with KeyCloak

### DISCUSSION THREAD

**Discussion by pdekraker-epa** [ckan/ckan#6188] [Q&A] [answered]
Our organization has setup Keycloak for identity management and I am looking to integrate it with our CKAN server running version 2.9. Looking around I see a few extensions that might help, but would be curious if the community has any suggestions.

[ckanext-keycloak](https://github.com/etri-odp/ckanext-keycloak) or [ckanext-sso](https://github.com/bcgov/ckanext-sso)
These seem to be the same, and have the most direct connection with Keycloak, but are out of date and I am unsure of the burden to get it up to Python 3 & CKAN 2.9+. 

[ckanext-saml2](https://github.com/DataShades/ckanext-saml2)
[ckanext-saml2auth](https://github.com/keitaroinc/ckanext-saml2auth)
I believe a SAML2 connection can be made to utilize Keycloak, but I am unsure what if anything I might be giving up by doing so?  

Are there suggestions for other solutions I am missing?

Thanks in advance...

**Comment by frafra** [comment]
It can happen to find outdated extensions. If you really need it, just port to CKAN 2.9, it is not so hard actually.

**Comment by krakhead22** [comment, marked as answer]
I ended up going with [ckanext-saml2auth](https://github.com/keitaroinc/ckanext-saml2auth) and it is working well for our usage.

**Reply by asdaqqaq** [reply]
Hello krakhead22 we need some help please,
we are using [ckanext-saml2auth] plugin but we have internal server error when we click on login button. we can not find a log that can help understand what is wrong, 
we defined saml2 in our red-hat SSO. and the URL for it is working. 
 these are the configs we used in ckan.ini file:

# Saml Configuration
ckanext.saml2auth.idp_metadata.location = remote
ckanext.saml2auth.idp_metadata.remote_url = https://xxx/auth/realms/ckan-dev/protocol/saml
ckanext.saml2auth.idp_metadata.remote_cert = /etc/pki/tls/certs/ckan-dev-client.crt
ckanext.saml2auth.user_firstname = firstname
ckanext.saml2auth.user_lastname = lastname
ckanext.saml2auth.user_fullname = fullname
ckanext.saml2auth.user_email = email

can you share the CKAN ini file config for the saml2auth?

ckanext.saml2auth.idp_metadata.remote_cert ----> is this the red-hat SSO certificate OR the saml2 client cert on SSO ?!

thank you in advance for your help..

**Reply by pdekraker-epa** [reply]
Looking at your config I think the remote url may need /descriptor appended

For the cert I the the saml2 client cert is the one you are looking for 

The internal server error is a little odd, have you confirmed it is thrown by the CKAN server? I also know there have been some recent changes to the login infrastructure in the master branch that could be related if that is what you are running.

**Reply by asdaqqaq** [reply]
thank you for your reply...


So we added the saml2 client cert then the error is thrown by ckan. We don't see anything in the logs. Also do we need to add the optional config mentioned here: https://github.com/keitaroinc/ckanext-saml2auth

Thanks!!!

**Reply by ghost** [reply]
Hello! I am running into issues when trying to implement SSO with KeyCloak and SAML. If any of you managed this, I would be glad to hear with what settings you make it work. I use CKAN 2.9.5.

**Comment by asdaqqaq** [comment]
We were not able to resolve this ☹ we appreciate any input!

Arwa Daqqaq, CEDA | Business Intelligence Specialist
Center for Enterprise Data & Analytics (CEDA)
Enterprise Business Intelligence

***@***.***<https://stateoftennessee.formstack.com/forms/sts_howsmyservice>

From: maxclac ***@***.***>
Sent: Tuesday, November 22, 2022 9:12 AM
To: ckan/ckan ***@***.***>
Cc: Arwa Daqqaq ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [ckan/ckan] User Authentication with KeyCloak (Discussion #6188)


*** This is an EXTERNAL email. Please exercise caution. DO NOT open attachments or click links from unknown senders or unexpected email - STS-Security. ***

Hello! I am running into issues when trying to implement SSO with KeyCloak and SAML. If any of you managed this, I would be glad to hear with what settings you make it work. I use CKAN 2.9.5.

**Comment by jqnatividad** [comment]
Just wanted to add that we recently forked ckanext-sso, and its working well for us for an enterprise implementation with very high security requirements that's powered by [AWS Cognito](https://aws.amazon.com/cognito/).

https://github.com/dathere/ckanext-sso
